lib/cl/cmdqueue.jl

```diff
@@ -6,7 +6,12 @@ mutable struct CmdQueue <: CLObject
     function CmdQueue(q_id::cl_command_queue; retain::Bool=false)
         q = new(q_id)
         retain && clRetainCommandQueue(q)
-        finalizer(clReleaseCommandQueue, q)
+        finalizer(q) do _
+          # this is to prevent `device_synchronize()` operating on freed queues.
+          # XXX: why does the WeakKeyDict contain freed objects?
```

Comment:
Order of  finalizers? E.g. the second finalizer that removes  `q` from queues hasn't run yet?

Reply:
https://github.com/JuliaLang/julia/blob/b9d9b69165493f6fc03870d975be05c67f14a30b/base/weakkeydict.jl#L93

Reply:
That, or maybe the objects are collected as part of a single collection cycle? In any case, kind-of surprising.